### ISSUE TITLE
Numba heapq doesn't work with numba typed lists

### ISSUE THREAD

**Issue by ma-sadeghi** [numba/numba#4926]
- [x] I am using the latest released version of Numba
- [x] I have included below a minimal working reproducer

# Description
`heapq` doesn't seem to like anything other than lists, even `numba` typed lists don't work.

# How to reproduce
```
import heapq as hq
from numba.typed import List

l = List()
l.append(5)
l.append(3)

e = hq.heappop(l)
```
```
TypeError: heap argument must be a list
```

**Comment by stuartarchibald** [comment]
Thanks for the report. The issue is that CPython requires a CPython list:
https://github.com/python/cpython/blob/664fe3996f7e05ae351526f02b21504bb065bcf8/Modules/_heapqmodule.c#L139-L142 pretty sure there's nothing that can easily be done about this from Numba.

Numba has it's own [`heapq` implementation](https://github.com/numba/numba/blob/7b2b0a1966b1d503e709a84da8aa64e72f35c9dc/numba/targets/heapq.py) for use in `jit` decorated functions, we should fix that to accept `typed.List`.

**Comment by ma-sadeghi** [comment]
@stuartarchibald Thank you for your quick reply. How can I use Numba's `heapq` implementation?

**Comment by stuartarchibald** [comment]
@ma-sadeghi there's a short example here: https://nbviewer.jupyter.org/github/numba/numba-examples/blob/master/notebooks/Numba_043_release_demo.ipynb#Support-for-heapq- essentially `import heapq` then use it in a jit decorated function, it behaves pretty much the same as CPython `heapq` (IIRC the implementations are very similar).

**Comment by gmarkall** [comment]
Bumped into this attempting to help @truenicoco with an implementation of a priority queue on Gitter. With the following PQ implementation:

```python
import itertools

import numba as nb
from numba.experimental import jitclass
from typing import List, Tuple, Dict
from heapq import heappush, heappop


# @jitclass
class PurePythonPriorityQueue:
    def __init__(self):
        self.pq = []  # list of entries arranged in a heap
        self.entry_finder = {}  # mapping of indices to entries
        self.REMOVED = -1  # placeholder for a removed item
        self.counter = itertools.count()  # unique sequence count

    def put(self, item: Tuple[int, int], priority: float = 0.0):
        """Add a new item or update the priority of an existing item"""
        if item in self.entry_finder:
            self.remove_item(item)
        count = next(self.counter)
        entry = [priority, count, item]
        self.entry_finder[item] = entry
        heappush(self.pq, entry)

    def remove_item(self, item: Tuple[int, int]):
        """Mark an existing item as REMOVED.  Raise KeyError if not found."""
        entry = self.entry_finder.pop(item)
        entry[-1] = self.REMOVED

    def pop(self):
        """Remove and return the lowest priority item. Raise KeyError if empty."""
        while self.pq:
            priority, count, item = heappop(self.pq)
            if item is not self.REMOVED:
                del self.entry_finder[item]
                return item
        raise KeyError("pop from an empty priority queue")


# @jitclass
# class Item:
#     i: int
#     j: int
#
#     def __init__(self, i, j):
#         self.i = i
#         self.j = j
#
#     def __eq__(self, other):
#         return self.i == other.i and self.j == other.j


@jitclass
class Entry:
    priority: float
    count: int
    item: Tuple[int, int]
    removed: bool

    def __init__(self, p: float, c: int, i: Tuple[int, int]):
        self.priority = p
        self.count = c
        self.item = i
        self.removed = False

    def __lt__(self, other):
        return self.priority < other.priority


@jitclass
class PriorityQueue:
    pq: List[Entry]
    entry_finder: Dict[Tuple[int, int], Entry]
    counter: int

    def __init__(self):
        self.pq = nb.typed.List.empty_list(Entry(0.0, 0, (0, 0)))
        self.entry_finder = nb.typed.Dict.empty((0, 0), Entry(0, 0, (0, 0)))
        self.counter = 0

    def put(self, item: Tuple[int, int], priority: float = 0.0):
        """Add a new item or update the priority of an existing item"""
        if item in self.entry_finder:
            self.remove_item(item)
        self.counter += 1
        entry = Entry(priority, self.counter, item)
        self.entry_finder[item] = entry
        heappush(self.pq, entry)

    def remove_item(self, item: Tuple[int, int]):
        """Mark an existing item as REMOVED.  Raise KeyError if not found."""
        entry = self.entry_finder.pop(item)
        entry.removed = True

    def pop(self):
        """Remove and return the lowest priority item. Raise KeyError if empty."""
        while self.pq:
            priority, count, item = heappop(self.pq)
            entry = heappop(self.pq)
            if not entry.removed:
                del self.entry_finder[entry.item]
                return item
        raise KeyError("pop from an empty priority queue")


if __name__ == "__main__":
    queue1 = PurePythonPriorityQueue()
    queue1.put((4, 5), 5.4)
    queue1.put((5, 6), 1.0)
    print(queue1.pop())  # Yay this works!

    queue2 = PriorityQueue()  # Nope
    queue2.put((4, 5), 5.4)
    queue2.put((5, 6), 1.0)
    print(queue2.pop())

    # queue3 = OptimalQueue()
    # for i in range(5):
    #     for j in range(5):
    #         queue3.add_index(i, j)
    #         # [Index(i, j) for i in range(5) for j in range(5)]
    # print(queue3.indices)
    # queue3.put((4, 5), 5.4)
    # queue3.put((5, 6), 1.0)
    # print(queue3.pop())  # Yay this works!
```

If I remove the check for a reflected list in the heapq implementation:

```diff
diff --git a/numba/cpython/heapq.py b/numba/cpython/heapq.py
index 23a80e269..50b1ddce3 100644
--- a/numba/cpython/heapq.py
+++ b/numba/cpython/heapq.py
@@ -104,8 +104,8 @@ def _heapreplace_max(heap, item):
 
 
 def assert_heap_type(heap):
-    if not isinstance(heap, types.List):
-        raise TypingError('heap argument must be a list')
+    #if not isinstance(heap, types.List):
+    #    raise TypingError('heap argument must be a list')
 
     dt = heap.dtype
     if isinstance(dt, types.Complex):
```

Then I get the traceback:

```
$ python modified.py :q
(5, 6)
Traceback (most recent call last):
  File "/home/gmarkall/numbadev/issues/so-priority-queue/modified.py", line 114, in <module>
    queue2.put((4, 5), 5.4)
  File "/home/gmarkall/numbadev/numba/numba/experimental/jitclass/boxing.py", line 59, in wrapper
    return method(*args, **kwargs)
  File "/home/gmarkall/numbadev/numba/numba/core/dispatcher.py", line 482, in _compile_for_args
    error_rewrite(e, 'typing')
  File "/home/gmarkall/numbadev/numba/numba/core/dispatcher.py", line 423, in error_rewrite
    raise e.with_traceback(None)
numba.core.errors.TypingError: Failed in nopython mode pipeline (step: nopython frontend)
- Resolution failure for literal arguments:
Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<built-in function heappush>) found for signature:

 >>> heappush(ListType[instance.jitclass.Entry#7f7c42dbee50<priority:float64,count:int64,item:UniTuple(int64 x 2),removed:bool>], instance.jitclass.Entry#7f7c42dbee50<priority:float64,count:int64,item:UniTuple(int64 x 2),removed:bool>)

There are 2 candidate implementations:
  - Of which 2 did not match due to:
  Overload in function 'heappush': File: numba/cpython/heapq.py: Line 150.
    With argument(s): '(ListType[instance.jitclass.Entry#7f7c42dbee50<priority:float64,count:int64,item:UniTuple(int64 x 2),removed:bool>], instance.jitclass.Entry#7f7c42dbee50<priority:float64,count:int64,item:UniTuple(int64 x 2),removed:bool>)':
   Rejected as the implementation raised a specific error:
     TypingError: Failed in nopython mode pipeline (step: nopython frontend)
   No implementation of function Function(<function _siftdown at 0x7f7c4294cee0>) found for signature:

    >>> _siftdown(ListType[instance.jitclass.Entry#7f7c42dbee50<priority:float64,count:int64,item:UniTuple(int64 x 2),removed:bool>], Literal[int](0), int64)

   There are 2 candidate implementations:
         - Of which 2 did not match due to:
         Overload in function 'register_jitable.<locals>.wrap.<locals>.ov_wrap': File: numba/core/extending.py: Line 159.
           With argument(s): '(ListType[instance.jitclass.Entry#7f7c42dbee50<priority:float64,count:int64,item:UniTuple(int64 x 2),removed:bool>], int64, int64)':
          Rejected as the implementation raised a specific error:
            TypingError: Failed in nopython mode pipeline (step: nopython frontend)
          No implementation of function Function(<built-in function lt>) found for signature:

           >>> lt(instance.jitclass.Entry#7f7c42dbee50<priority:float64,count:int64,item:UniTuple(int64 x 2),removed:bool>, instance.jitclass.Entry#7f7c42dbee50<priority:float64,count:int64,item:UniTuple(int64 x 2),removed:bool>)

          There are 20 candidate implementations:
                - Of which 18 did not match due to:
                Overload of function 'lt': File: <numerous>: Line N/A.
                  With argument(s): '(instance.jitclass.Entry#7f7c42dbee50<priority:float64,count:int64,item:UniTuple(int64 x 2),removed:bool>, instance.jitclass.Entry#7f7c42dbee50<priority:float64,count:int64,item:UniTuple(int64 x 2),removed:bool>)':
                 No match.
                - Of which 2 did not match due to:
                Operator Overload in function 'lt': File: unknown: Line unknown.
                  With argument(s): '(instance.jitclass.Entry#7f7c42dbee50<priority:float64,count:int64,item:UniTuple(int64 x 2),removed:bool>, instance.jitclass.Entry#7f7c42dbee50<priority:float64,count:int64,item:UniTuple(int64 x 2),removed:bool>)':
                 No match for registered cases:
                  * (bool, bool) -> bool
                  * (int8, int8) -> bool
                  * (int16, int16) -> bool
                  * (int32, int32) -> bool
                  * (int64, int64) -> bool
                  * (uint8, uint8) -> bool
                  * (uint16, uint16) -> bool
                  * (uint32, uint32) -> bool
                  * (uint64, uint64) -> bool
                  * (float32, float32) -> bool
                  * (float64, float64) -> bool
          
          During: typing of intrinsic-call at /home/gmarkall/numbadev/numba/numba/cpython/heapq.py (18)
          
          File "../../numba/numba/cpython/heapq.py", line 18:
          def _siftdown(heap, startpos, pos):
              <source elided>
                  parent = heap[parentpos]
                  if newitem < parent:
                  ^
   
     raised from /home/gmarkall/numbadev/numba/numba/core/typeinfer.py:1074
   
   During: resolving callee type: Function(<function _siftdown at 0x7f7c4294cee0>)
   During: typing of call at /home/gmarkall/numbadev/numba/numba/cpython/heapq.py (157)
   
   
   File "../../numba/numba/cpython/heapq.py", line 157:
       def hq_heappush_impl(heap, item):
           <source elided>
           heap.append(item)
           _siftdown(heap, 0, len(heap) - 1)
           ^

  raised from /home/gmarkall/numbadev/numba/numba/core/typeinfer.py:1074

During: resolving callee type: Function(<built-in function heappush>)
During: typing of call at /home/gmarkall/numbadev/issues/so-priority-queue/modified.py (89)


File "modified.py", line 89:
    def put(self, item: Tuple[int, int], priority: float = 0.0):
        <source elided>
        self.entry_finder[item] = entry
        heappush(self.pq, entry)
        ^

- Resolution failure for non-literal arguments:
None

During: resolving callee type: BoundFunction((<class 'numba.core.types.misc.ClassInstanceType'>, 'put') for instance.jitclass.PriorityQueue#7f7c42d484f0<pq:ListType[instance.jitclass.Entry#7f7c42dbee50<priority:float64,count:int64,item:UniTuple(int64 x 2),removed:bool>],entry_finder:DictType[UniTuple(int64 x 2),instance.jitclass.Entry#7f7c42dbee50<priority:float64,count:int64,item:UniTuple(int64 x 2),removed:bool>]<iv=None>,counter:int64>)
During: typing of call at <string> (3)


File "<string>", line 3:
<source missing, REPL/exec in use?>
```

It seems like this comes from line 18 of heapq.py, `if newitem < parent:`:

```python
 11 @register_jitable
 12 def _siftdown(heap, startpos, pos):
 13     newitem = heap[pos]
 14 
 15     while pos > startpos:
 16         parentpos = (pos - 1) >> 1
 17         parent = heap[parentpos]
 18         if newitem < parent:
 19             heap[pos] = parent
 20             pos = parentpos
 21             continue
 22         break
 23 
 24     heap[pos] = newitem
```

**Comment by gmarkall** [comment]
Fix proposed in #7260.
